src/faebryk/libs/picker/jlcpcb/jlcpcb.py

```diff
@@ -0,0 +1,604 @@
+                f"'{self.extra['attributes'][attr_key]}', Error: '{e}'"
+            )
+
+        return F.ANY()
+
+    def attach(
+        self,
+        module: Module,
+        mapping: list[MappingParameterDB],
+        qty: int = 1,
+    ):
+        params = [
+            self.get_parameter(
+                attribute_search_keys=m.attr_keys,
+                tolerance_search_key=m.attr_tolerance_key,
+                parser=m.transform_fn,
+            )
+            for m in mapping
+        ]
+
+        for name, value in zip([m.param_name for m in mapping], params):
+            module.PARAMs.__getattribute__(name).merge(value)
+
+        F.has_defined_descriptive_properties.add_properties_to(
+            module,
+            {
+                DescriptiveProperties.partno: self.mfr,
+                DescriptiveProperties.manufacturer: asyncio.run(
+                    Manufacturers().get_from_id(self.manufacturer_id)
+                ),
+                DescriptiveProperties.datasheet: self.datasheet,
+                "JLCPCB stock": str(self.stock),
+                "JLCPCB price": f"{self.get_price(qty):.4f}",
+                "JLCPCB description": self.description,
+                "JLCPCB Basic": str(bool(self.basic)),
+                "JLCPCB Preferred": str(bool(self.preferred)),
+            },
+        )
+
+        module.add_trait(has_part_picked_defined(JLCPCB_Part(f"C{self.lcsc}")))
+
+        if not module.has_trait(F.can_attach_to_footprint):
+            pinmap = auto_pinmapping(module, f"C{self.lcsc}")
+            assert pinmap is not None
+            module.add_trait(F.can_attach_to_footprint_via_pinmap(pinmap))
+
+        attach_footprint(module, f"C{self.lcsc}", True)
+
+
+class ComponentQuery:
+    def __init__(self):
+        self.Q = Q()
+        self.results: list[Component] = []
+
+    async def exec(self) -> list[Component]:
+        self.results = await Component.filter(self.Q)
+        return self.results
+
+    def get(self) -> list[Component]:
+        return self.results or asyncio.run(self.exec())
+
+    def by_stock(self, qty: int) -> Self:
+        assert not self.results
+        self.Q &= Q(stock__gte=qty)
+        return self
+
+    def by_value(self, value: Parameter, si_unit: str) -> Self:
+        assert not self.results
+        value_query = Q()
+        for r in e_series_intersect(
+            value.get_most_narrow(), E_SERIES_VALUES.E_ALL
+        ).params:
+            assert isinstance(r, F.Constant)
+            si_val = float_to_si_str(r.value, si_unit).replace("µ", "u")
+            value_query |= Q(description__contains=f" {si_val}")
+        self.Q &= value_query
+        return self
+
+    def by_category(self, category: str, subcategory: str) -> Self:
+        assert not self.results
+        category_ids = asyncio.run(Category().get_ids(category, subcategory))
+        self.Q &= Q(category_id__in=category_ids)
+        return self
```

Comment:
I would call all the ComponentQuery functions filter_by_... (if they filter ofc)